staging/src/k8s.io/apiserver/pkg/admission/plugin/validatingadmissionpolicy/validator.go

```diff
@@ -18,298 +18,85 @@ package validatingadmissionpolicy
-	for i, validation := range p.Spec.Validations {
-		compilationResults[i] = CompileValidatingPolicyExpression(validation.Expression, hasParam)
-	}
-	return &CELValidator{policy: p, compilationResults: compilationResults}
-}
-
-// CELValidator implements the Validator interface
-type CELValidator struct {
-	policy             *v1alpha1.ValidatingAdmissionPolicy
-	compilationResults []CompilationResult
 }
 
-func convertObjectToUnstructured(obj interface{}) (*unstructured.Unstructured, error) {
-	if obj == nil || reflect.ValueOf(obj).IsNil() {
-		return &unstructured.Unstructured{Object: nil}, nil
-	}
-	ret, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
-	if err != nil {
-		return nil, err
+func policyDecisionActionForError(f v1.FailurePolicyType) cel.PolicyDecisionAction {
+	if f == v1.Ignore {
+		return cel.ActionAdmit
 	}
-	return &unstructured.Unstructured{Object: ret}, nil
+	return cel.ActionDeny
 }
 
-func objectToResolveVal(r runtime.Object) (interface{}, error) {
-	if r == nil || reflect.ValueOf(r).IsNil() {
-		return nil, nil
-	}
-	v, err := convertObjectToUnstructured(r)
-	if err != nil {
-		return nil, err
-	}
-	return v.Object, nil
-}
-
-func policyDecisionActionForError(f v1alpha1.FailurePolicyType) PolicyDecisionAction {
-	if f == v1alpha1.Ignore {
-		return ActionAdmit
-	}
-	return ActionDeny
-}
-
-// Validate validates all cel expressions in Validator and returns a PolicyDecision for each CEL expression or returns an error.
-// An error will be returned if failed to convert the object/oldObject/params/request to unstructured.
-// Each PolicyDecision will have a decision and a message.
-// policyDecision.message will be empty if the decision is allowed and no error met.
-func (v *CELValidator) Validate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) ([]PolicyDecision, error) {
-	// TODO: replace unstructured with ref.Val for CEL variables when native type support is available
-	decisions := make([]PolicyDecision, len(v.compilationResults))
-	var err error
-
-	oldObjectVal, err := objectToResolveVal(versionedAttr.VersionedOldObject)
-	if err != nil {
-		return nil, err
-	}
-	objectVal, err := objectToResolveVal(versionedAttr.VersionedObject)
-	if err != nil {
-		return nil, err
-	}
-	paramsVal, err := objectToResolveVal(versionedParams)
-	if err != nil {
-		return nil, err
+// Validate takes a list of Evaluation and a failure policy and converts them into actionable PolicyDecisions
+func (v *validator) Validate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) []cel.PolicyDecision {
+	var f v1.FailurePolicyType
+	if v.failPolicy == nil {
+		f = v1.Fail
+	} else {
+		f = *v.failPolicy
 	}
 
-	request := createAdmissionRequest(versionedAttr.Attributes)
-	requestVal, err := convertObjectToUnstructured(request)
+	evalResults, err := v.filter.ForInput(versionedAttr, versionedParams, cel.CreateAdmissionRequest(versionedAttr.Attributes))
```

Comment:
Where is CompilationErrors be handled?

Reply:
Currently this controller does not write out any status information. Errors are returned as rejections (subject to failure policy). We should report status on compilation error, since the target audience for such errors are policy authors/cluster admins, not the user making requests to the API server.

Adding status is out-of-scope for this PR (which maintains current behavior), but having access to compilation errors unblocks that work.